Several fixes and small enhancements.

f0rki opened the pull request:
let me know if you want them split into separate PRs.

hgarrereyn commented on core/graph.hpp:
```diff
@@ -348,11 +348,13 @@ class TGraph {
-        while (sample >= 0) {
-            sample_idx += 1;
+        unsigned int sample_idx = 0;
+        while (sample >= 0 && sample_idx < tree.children.size()) {
```
looks like `sample_idx` needs to update inside the loop here

hgarrereyn commented on core/graph.hpp:
```diff
@@ -348,11 +348,13 @@ class TGraph {
+        while (sample >= 0 && sample_idx < tree.children.size()) {
             sample -= tree.children[sample_idx].num_subtrees;
         }
+        if (sample_idx >= tree.children.size()) {
```
curious if you hit cases where this went out of bounds; if so, may be a bug with the `num_subtrees` computation

f0rki replied:
yes. I observed a OOB on `tree.childern[sample_idx]`.

f0rki replied:
on second thought. maybe this was just an attempt to fix the loop above.

hgarrereyn commented on core/harness.cpp:
```diff
@@ -426,12 +437,17 @@ extern "C" int LLVMFuzzerTestOneInput(const uint8_t *Data, size_t Size) {
     for (Node n : nodes) {
-        void *in_ref[n.in_ref_size()];
-        void *out_ref[n.out_ref_size()];
+        void *in_ref[n.in_ref_size() + 1];
```
could you explain this part

f0rki replied:
UBSan detects zero sized VLA on the stack as UB, so I just made sure the array has at least one element, even if it is not used.

hgarrereyn commented on cli/gfuzz/commands/gen/cpp/gen_cpp.py:
```diff
@@ -53,7 +53,7 @@
 
 CPP_SAVE_REF = '''\
-    out_ref[{out_ref_idx}] = reinterpret_cast<void *>({arg_name});
+    out_ref[{out_ref_idx}] = (void*)({arg_name});
```
I think there was a reason I used `reinterpret_cast` here over C-style casts, bit I can't remember the specifics. Can you explain the rationale for this change?

f0rki replied:
I had troubles with harnessing a C library unless I used this. (because of casting a const pointer to a non-const pointer and vice-versa or something. I think const to non-const cast is forbidden with `reinterpret_cast`).

hgarrereyn reviewed (commented):
Hey, thanks for the code! Really appreciate the schema validation you added; I think that would be a big help for people writing harnesses. I left some comments/questions on the details. Also, regarding the ASAN build and some of the uninit protections you added to core, were you hitting errors caused by uninitialized reads or is this just proactive?

f0rki commented:
ok let me redo the bugfix. There might be some unnecessary changes.

f0rki committed:
added install script for easy and quick installation

f0rki committed:
gfuzz generator: added some more schema validation to catch obvious errors such as typos in the schema

f0rki committed:
Fixed potential OOB access in the graph handling part of the C++ fuzzing code.

f0rki committed:
Better handling of the initializer and finalizer code pieces.

* `global_init` is now a configurable piece of code specified in the
  schema. This allows to perform some setup once globally vs. at every
  testcase (`initializer` and `finalizer` are called on every `LLVMFuzzerTestOneInput`)
* Output the code pieces from `global_init`, `initializer`, and `finalizer` in the testcase writer.

f0rki committed:
use C-style cast instead of reinterpret_cast to get rid of const warnings

f0rki committed:
Nicer printing of char arrays in the writer.

f0rki commented:
> were you hitting errors caused by uninitialized reads or is this just proactive?

I believe this might have been caused by the issue with the subtree sampling in `AppendTree`. I switched to do this only if debug flag is set.
